src/token/TokenSwap.sol

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity >=0.8.15;
+import "./TransferHelper.sol";
+import "../access/AccessControl.sol";
+import "../utils/Cast.sol";
+
+
+/// @dev TokenSwap is a contract that can be used swap tokens at a fixed rate, with
+/// the aim of completely replacing the supply of a token by the funds supplied to
+/// this contract. It is meant to be used as a token upgrade, when other mechanisms fail.
+contract TokenSwap is AccessControl() {
+    using Cast for uint256;
+    using TransferHelper for IERC20;
+
+    event Registered(IERC20 indexed tokenIn, IERC20 indexed tokenOut, uint256 tokenInBalance, uint256 tokenOutBalance, uint96 ratio);
+    event Unregistered(IERC20 indexed tokenIn, IERC20 indexed tokenOut, uint256 tokenInBalance, uint256 tokenOutBalance);
+    event Swapped(IERC20 indexed tokenIn, IERC20 indexed tokenOut, uint256 tokenInAmount, uint256 tokenOutAmount);
+    event Extracted(IERC20 indexed tokenIn, uint256 tokenInBalance);
+    event Recovered(IERC20 indexed token, uint256 recovered);
+
+    struct TokenIn {
+        IERC20 reverse;
+        uint96 ratio;
+        uint256 balance;
+    }
+
+    struct TokenOut {
+        IERC20 reverse;
+        uint256 balance;     
+    }
+
+    mapping (IERC20 => TokenIn) public tokensIn;
+    mapping (IERC20 => TokenOut) public tokensOut;
+
+    /// @dev Register a token to be replaced, and the token to replace it with.
+    /// The ratio is calculated as the funds of the replacement token divided by the supply of the token to be replaced.
+    /// The tokens used as a replacement must have been sent to the contract before this call.
+    /// @param tokenIn_ The token to be replaced
+    /// @param tokenOut_ The token to replace it with
+    function register(IERC20 tokenIn_, IERC20 tokenOut_) external auth {
+        require(address(tokenIn_) != address(tokenOut_), "Same token");
+        require(address(tokensIn[tokenIn_].reverse) == address(0), "TokenIn already registered");
+        require(address(tokensOut[tokenOut_].reverse) == address(0), "TokenOut already registered");
```

Comment:
Should each `tokenIn_` to `tokenOut_` have a 1-1 relationship? ie should you only be able to swap token a *only* for token b?

Reply:
As an example, this doesn't look like it would allow for say USDC to be set as the `tokenIn_` with both DAI and FRAX as `tokenOut_`. Is that desired behaviour?